radicle-tracker/src/thread.rs

```diff
@@ -0,0 +1,827 @@
+    ///     ReplyTo::Thread
+    /// );
+    ///
+    /// thread.reply(String::from("What should we use them for?"), ReplyTo::Main);
+    /// thread.edit(|body| *body = String::from("How can we use them?"));
+    ///
+    /// assert_eq!(thread.view(), Ok(&Status::Live(String::from("How can we use them?"))));
+    /// #
+    /// #     Ok(())
+    /// # }
+    /// ```
+    pub fn edit<F>(&mut self, f: F) -> Result<(), Error>
+    where
+        F: FnOnce(&mut A) -> (),
+    {
+        match self._finger {
+            Finger::Root => {
+                f(self.root.get_mut());
+                Ok(())
+            },
+            Finger::Main(main_ix) => {
+                let node = self.index_main_mut(main_ix).first_mut();
+                f(node.get_mut());
+                Ok(())
+            },
+            Finger::Thread((main_ix, replies_ix)) => {
+                let replies = self.index_main_mut(main_ix);
+                let node = replies
+                    .get_mut(replies_ix)
+                    .unwrap_or_else(|| panic!("Reply index is out of bounds: {}", replies_ix));
+                f(node.get_mut());
+                Ok(())
+            },
+        }
+    }
+
+    /// Expand the current main thread item we are looking at into the full
+    /// non-empty view of items.
+    ///
+    /// # Panics
+    ///
+    /// If the internal finger into the thread is out of bounds.
+    pub fn expand(&self) -> NonEmpty<Status<A>>
+    where
+        A: Clone,
+    {
+        let main_ix = match self._finger {
+            Finger::Root => {
+                return NonEmpty::from((
+                    self.root.clone(),
+                    self.main_thread
+                        .clone()
+                        .iter()
+                        .map(|thread| thread.first().clone())
+                        .collect(),
+                ));
+            },
+            Finger::Main(main_ix) => main_ix,
+            Finger::Thread((main_ix, _)) => main_ix,
+        };
+
+        self.index_main(main_ix).0.clone()
+    }
+
+    /// Look at the current item we are pointing to in the thread.
+    ///
+    /// # Panics
+    ///
+    /// If the internal finger into the thread is out of bounds.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use radicle_tracker::{Status, Thread};
+    ///
+    /// let (thread) = Thread::new(String::from("Discussing rose trees"));
+    ///
+    /// assert_eq!(thread.view(), Ok(&Status::Live(String::from("Discussing rose trees"))));
+    /// ```
+    pub fn view(&self) -> Result<&Status<A>, Error> {
+        match self._finger {
+            Finger::Root => Ok(&self.root),
+            Finger::Main(main_ix) => Ok(self.index_main(main_ix).first()),
+            Finger::Thread((main_ix, replies_ix)) => Ok(self
+                .index_main(main_ix)
+                .get(replies_ix)
+                .unwrap_or_else(|| panic!("Reply index is out of bounds: {}", replies_ix))),
```

Comment:
That's kind of like `expect(format!("{}", ...))` (yes it's slightly less efficient but maybe more concise)

Reply:
Clippy will give out to you for that because of evaluation. It's what I had originally :sweat_smile:

Reply:
hah I see.